// ungit/components/modals/modals.bundle.js
(function(){function r(e,n,t){function o(i,f){if(!n[i]){if(!e[i]){var c="function"==typeof require&&require;if(!f&&c)return c(i,!0);if(u)return u(i,!0);var a=new Error("Cannot find module '"+i+"'");throw a.code="MODULE_NOT_FOUND",a}var p=n[i]={exports:{}};e[i][0].call(p.exports,function(r){var n=e[i][1][r];return o(n||r)},p,p.exports,r,e,n,t)}return n[i].exports}for(var u="function"==typeof require&&require,i=0;i<t.length;i++)o(t[i]);return o}return r})()({1:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var ko = require("knockout");
var modalBase_1 = require("./modalBase");
ungit.components.register('credentialsmodal', function (args) { return new CredentialsModalViewModel(args.remote); });
ungit.components.register('addremotemodal', function (arg) { return new AddRemoteModalViewModel(arg.path); });
ungit.components.register('addsubmodulemodal', function (arg) { return new AddSubmoduleModalViewModel(arg.path); });
/**
 * Form receives collection of user inputs, i.e. username, password and etc.
 */
var FormModalViewModel = /** @class */ (function (_super) {
    __extends(FormModalViewModel, _super);
    function FormModalViewModel(title, taModalName, showCancel) {
        var _this = _super.call(this, title, taModalName) || this;
        _this.items = [];
        _this.showCancel = showCancel;
        _this.template = 'formModal';
        return _this;
    }
    FormModalViewModel.prototype.submit = function () {
        this.close();
    };
    return FormModalViewModel;
}(modalBase_1.ModalViewModel));
var CredentialsModalViewModel = /** @class */ (function (_super) {
    __extends(CredentialsModalViewModel, _super);
    function CredentialsModalViewModel(remote) {
        var _this = _super.call(this, "Remote ".concat(remote, " requires authentication"), 'credentials-dialog', false) || this;
        _this.items.push(new modalBase_1.FormItems('Username', ko.observable(), 'text', true));
        _this.items.push(new modalBase_1.FormItems('Password', ko.observable(), 'password', false));
        return _this;
    }
    CredentialsModalViewModel.prototype.submit = function () {
        _super.prototype.submit.call(this);
        ungit.programEvents.dispatch({
            event: 'request-credentials-response',
            username: this.items[0].value(),
            password: this.items[1].value()
        });
    };
    return CredentialsModalViewModel;
}(FormModalViewModel));
var AddRemoteModalViewModel = /** @class */ (function (_super) {
    __extends(AddRemoteModalViewModel, _super);
    function AddRemoteModalViewModel(path) {
        var _this = _super.call(this, 'Add new remote', 'add-remote', true) || this;
        _this.repoPath = path;
        _this.items.push(new modalBase_1.FormItems('Name', ko.observable(), 'text', true));
        _this.items.push(new modalBase_1.FormItems('Url', ko.observable(), 'text', false));
        return _this;
    }
    AddRemoteModalViewModel.prototype.submit = function () {
        return __awaiter(this, void 0, void 0, function () {
            var e_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _super.prototype.submit.call(this);
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, ungit.server.postPromise("/remotes/".concat(encodeURIComponent(this.items[0].value())), {
                                path: this.repoPath,
                                url: this.items[1].value()
                            })];
                    case 2:
                        _a.sent();
                        ungit.programEvents.dispatch({ event: 'update-remote' });
                        return [3 /*break*/, 4];
                    case 3:
                        e_1 = _a.sent();
                        ungit.server.unhandledRejection(e_1);
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    return AddRemoteModalViewModel;
}(FormModalViewModel));
var AddSubmoduleModalViewModel = /** @class */ (function (_super) {
    __extends(AddSubmoduleModalViewModel, _super);
    function AddSubmoduleModalViewModel(path) {
        var _this = _super.call(this, 'Add new submodule', 'add-submodule', true) || this;
        _this.repoPath = path;
        _this.items.push(new modalBase_1.FormItems('Path', ko.observable(), 'text', true));
        _this.items.push(new modalBase_1.FormItems('Url', ko.observable(), 'text', false));
        return _this;
    }
    AddSubmoduleModalViewModel.prototype.submit = function () {
        return __awaiter(this, void 0, void 0, function () {
            var e_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _super.prototype.submit.call(this);
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, ungit.server.postPromise('/submodules/add', {
                                path: this.repoPath,
                                submodulePath: this.items[0].value(),
                                submoduleUrl: this.items[1].value()
                            })];
                    case 2:
                        _a.sent();
                        ungit.programEvents.dispatch({ event: 'submodule-fetch' });
                        return [3 /*break*/, 4];
                    case 3:
                        e_2 = _a.sent();
                        ungit.server.unhandledRejection(e_2);
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    return AddSubmoduleModalViewModel;
}(FormModalViewModel));

},{"./modalBase":2,"knockout":undefined}],2:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromptOptions = exports.FormItems = exports.ModalViewModel = void 0;
var ModalViewModel = /** @class */ (function () {
    function ModalViewModel(title, taModalName) {
        this.timestamp = new Date().getTime();
        this.title = title;
        this.taModalName = taModalName;
    }
    ModalViewModel.prototype.close = function () {
        ungit.programEvents.dispatch({ event: 'modal-close-dialog', modal: this });
    };
    return ModalViewModel;
}());
exports.ModalViewModel = ModalViewModel;
var FormItems = /** @class */ (function () {
    function FormItems(name, value, type, autoFocus) {
        this.name = name;
        this.value = value;
        this.type = type;
        this.autoFocus = autoFocus;
    }
    return FormItems;
}());
exports.FormItems = FormItems;
var PromptOptions = /** @class */ (function () {
    function PromptOptions(label, primary, taId, close) {
        this.label = label;
        this.primary = primary;
        this.taId = taId;
        this.close = close;
    }
    return PromptOptions;
}());
exports.PromptOptions = PromptOptions;

},{}],3:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("./forms");
require("./prompts");

},{"./forms":1,"./prompts":4}],4:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var ko = require("knockout");
var modalBase_1 = require("./modalBase");
ungit.components.register('yesnomodal', function (args) { return new YesNoModalViewModel(args.title, args.details, args.closeFunc); });
ungit.components.register('yesnomutemodal', function (args) { return new YesNoMuteModalViewModel(args.title, args.details, args.closeFunc); });
ungit.components.register('toomanyfilesmodal', function (args) { return new TooManyFilesModalViewModel(args.title, args.details, args.closeFunc); });
ungit.components.register('texteditmodal', function (args) { return new TextEditModal(args.title, args.content, args.closeFunc); });
/**
 * Prompt's receives decisions from users, such as 'yes' or 'no', based on
 * button clicks.
 */
var PromptModalViewModel = /** @class */ (function (_super) {
    __extends(PromptModalViewModel, _super);
    function PromptModalViewModel(title, taModalName, details, closeFunc) {
        var _this = _super.call(this, title, taModalName) || this;
        _this.promptOptions = [];
        _this.details = ko.observable(details);
        _this.template = 'promptModal';
        _this.closeFunc = closeFunc;
        return _this;
    }
    PromptModalViewModel.prototype.close = function (isYes, isMute) {
        if (isYes === void 0) { isYes = false; }
        if (isMute === void 0) { isMute = false; }
        this.closeFunc(isYes, isMute);
        _super.prototype.close.call(this);
    };
    PromptModalViewModel.prototype.closeYes = function () {
        this.close(true);
    };
    PromptModalViewModel.prototype.closeYesMute = function () {
        this.close(true, true);
    };
    PromptModalViewModel.prototype.closeNo = function () {
        this.close();
    };
    return PromptModalViewModel;
}(modalBase_1.ModalViewModel));
var YesNoModalViewModel = /** @class */ (function (_super) {
    __extends(YesNoModalViewModel, _super);
    function YesNoModalViewModel(title, details, closeFunc) {
        var _this = _super.call(this, title, 'yes-no-modal', details, closeFunc) || this;
        _this.promptOptions.push(new modalBase_1.PromptOptions('Yes', true, 'yes', _this.closeYes.bind(_this)));
        _this.promptOptions.push(new modalBase_1.PromptOptions('No', false, 'no', _this.closeNo.bind(_this)));
        return _this;
    }
    return YesNoModalViewModel;
}(PromptModalViewModel));
var YesNoMuteModalViewModel = /** @class */ (function (_super) {
    __extends(YesNoMuteModalViewModel, _super);
    function YesNoMuteModalViewModel(title, details, closeFunc) {
        var _this = _super.call(this, title, 'yes-no-mute-modal', details, closeFunc) || this;
        _this.promptOptions.push(new modalBase_1.PromptOptions('Yes', true, 'yes', _this.closeYes.bind(_this)));
        _this.promptOptions.push(new modalBase_1.PromptOptions('Yes and mute for awhile', false, 'mute', _this.closeYesMute.bind(_this)));
        _this.promptOptions.push(new modalBase_1.PromptOptions('No', false, 'no', _this.closeNo.bind(_this)));
        return _this;
    }
    return YesNoMuteModalViewModel;
}(PromptModalViewModel));
var TooManyFilesModalViewModel = /** @class */ (function (_super) {
    __extends(TooManyFilesModalViewModel, _super);
    function TooManyFilesModalViewModel(title, details, closeFunc) {
        var _this = _super.call(this, title, 'yes-no-modal', details, closeFunc) || this;
        _this.promptOptions.push(new modalBase_1.PromptOptions("Don't load", true, 'noLoad', _this.closeYes.bind(_this)));
        _this.promptOptions.push(new modalBase_1.PromptOptions("Load anyway", false, 'loadAnyway', _this.closeNo.bind(_this)));
        return _this;
    }
    return TooManyFilesModalViewModel;
}(PromptModalViewModel));
var TextEditModal = /** @class */ (function (_super) {
    __extends(TextEditModal, _super);
    function TextEditModal(title, details, closeFunc) {
        var _this = _super.call(this, title, 'text-edit-modal', "<textarea class=\"text-area-content form-control\" spellcheck=\"false\" style=\"height: 250px; width: 100%; font-family: monospace; resize: vertical;\">".concat(details, "</textarea>"), closeFunc) || this;
        _this.promptOptions.push(new modalBase_1.PromptOptions('Save', true, 'save', _this.closeYes.bind(_this)));
        _this.promptOptions.push(new modalBase_1.PromptOptions('Cancel', false, 'cancel', _this.closeNo.bind(_this)));
        return _this;
    }
    return TextEditModal;
}(PromptModalViewModel));

},{"./modalBase":2,"knockout":undefined}]},{},[3])
